### PR TITLE
Sum_list

### LINKED ISSUE #76: sum_list

### PR DESCRIPTION
**Pull request by Jha-mal**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/sum_list.py
Comment (ermishina) [inline solutions/sum_list.py:1]: The code looks structured and is accompanied by explanatory docstrings. Defensive checks are implemented correctly. The function uses built-in Python tools (sum), which makes it concise and performant.
Examples in the docstring illustrate function usage effectively, which is particularly helpful for users and maintainers.
The use of type hints (List[float], -> float) improves code readability and ensures type safety.

File: solutions/sum_list.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+Doctests:
+    >>> sum_list([1, 2, 3])
+    6
+    >>> sum_list([-1, -2, -3])
+    -6
+    >>> sum_list([0, 0, 1.5])
+    1.5
+    >>> sum_list([])
+    0
+"""
```
Comment (ermishina) [inline solutions/sum_list.py:33]: The docstring is well written, but you could add an example with a larger list to show how the function works with edge cases.

File: solutions/tests/test_sum_list.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        Test that sum_list correctly sums negative numbers.
+        """
+        self.assertEqual(sum_list([-1, -2, -3]), -6)
+
+    def test_sums_mixed_numbers(self):
+        """
+        Test that sum_list correctly sums a mix of negative, positive, and floats.
+        """
+        self.assertEqual(sum_list([-1, 1, 2.5]), 2.5)
+
```
Comment (ermishina) [inline solutions/tests/test_sum_list.py:42]: The test test_sums_mixed_numbers uses a floating-point number (2.5). Summing floats can result in precision errors, so it’s better to use assertAlmostEqual:
self.assertAlmostEqual(sum_list([-1, 1, 2.5]), 2.5)

File: solutions/tests/test_sum_list.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        """
+        self.assertEqual(sum_list([]), 0)
+
+    def test_defensive_check_non_numeric(self):
+        """
+        Test that a TypeError is raised if non-numeric items are included.
+        """
+        with self.assertRaises(TypeError):
+            sum_list([1, "two", 3])
+
```
Comment (ermishina) [inline solutions/tests/test_sum_list.py:55]: The test for invalid data (test_defensive_check_non_numeric) verifies that a TypeError is raised. However, it might be unclear if a different exception type is raised. You can add a message for clarification:
with self.assertRaises(TypeError, msg="sum_list should raise TypeError for invalid data"):

File: solutions/tests/test_sum_list.py
Comment (ermishina) [inline solutions/tests/test_sum_list.py:1]: Various scenarios are covered: positive numbers, negative numbers, mixed types, an empty list, and handling invalid data. Each test has a clear purpose.
The code is well-structured: each test covers a specific scenario.
The behavior and purpose of the tests are well-documented.

You can improve:
A comment or example could be added to handle edge cases like None instead of an empty list.
While the test coverage is solid, adding a case for very large lists or numbers could ensure the function handles such scenarios effectively.

File: solutions/sum_list.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      >>> sum_list([-1, -2.5, 3])
+      -0.5
+      >>> sum_list([])
+      0
+    """
+
+    # Defensive assertion (checks exactly one assumption: item must be numeric)
+    for item in numbers:
+        if not isinstance(item, (int, float)):
+            raise TypeError("All items must be numeric (int or float).")
```
Comment (ermishina) [inline solutions/sum_list.py:70]: The error message in the TypeError could be expanded for better user clarity, e.g.:
raise TypeError("All items in the list must be numeric (int or float). Found a non-numeric value.")

File: solutions/sum_list.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      :return: float
+        The sum of all items in 'numbers' as a float.
+
+    Raises:
+      TypeError:
+        If any item in 'numbers' is not numeric (int or float).
+
+    :examples:
+      >>> sum_list([1.0, 2.0, 3.0])
+      6.0
```
Comment (ermishina) [inline solutions/sum_list.py:60]: The doctest uses floating-point numbers ([1.0, 2.0, 3.0]), but the test cases in the related file include integers. Consider aligning the examples for consistency.

**Review by ermishina — APPROVED** [review]
Tests cover key cases. Minor improvements, like optimizing data validation and clarifying error messages, could enhance it further. Overall, it's a solid and efficient implementation.

**Review by ermishina — APPROVED** [review]
